contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/gst_v4l2_audio_video_synchronization.py

```diff
@@ -0,0 +1,188 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Checkbox. If not, see <http://www.gnu.org/licenses/>.
+
+import argparse
+import logging
+import os
+import shlex
+import subprocess
+from typing import Any
+
+logging.basicConfig(level=logging.INFO)
+
+
+def register_arguments():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description=(
+            "Script helps playing a golden sample video on specific display"
+            ' by the specific "video sink", e.g. waylandsink. The golden '
+            "is special video for verifying AV Synchronization."
+        ),
+    )
+
+    parser.add_argument(
+        "-gp",
+        "--golden_sample_path",
+        required=True,
+        type=str,
+        help="Path of Golden Sample file",
+    )
+
+    parser.add_argument(
+        "-dp",
+        "--decoder_plugin",
+        required=True,
+        type=str,
+        help="Decoder plugin be used in gstreamer pipeline e.g. v4l2h264dec",
+    )
+
+    parser.add_argument(
+        "-vs",
+        "--video_sink",
+        default="waylandsink",
+        type=str,
+        help=(
+            "Specific value of video-sink for gstreamer that a video can be"
+            " displayed on. (Default: waylandsink)"
+        ),
+    )
+
+    args = parser.parse_args()
+    return args
+
+
+def build_gst_command(
+    gst_bin: str, golden_sample_path: str, decoder: str, video_sink: str
+) -> str:
+    """
+    Builds a GStreamer command to process the golden sample.
+
+    :param gst_bin:
+        The binary name of gstreamer. Default is "gst-launch-1.0"
+        You can assign the snap name to GST_LAUNCH_BIN env variable if you
+        want to using snap.
+    :param golden_sample:
+        The path to the golden sample file.
+    :param decoder:
+        The decoder to use for the video, e.g., "v4l2vp8dec", "v4l2vp9dec".
+    :param video_sink:
+        The specific sink for video displaying on, e.g. "waylandsink"
+
+    :returns:
+        The GStreamer command to execute.
+    """
+
+    # FIXME: Why we did this?
+    # Because the 4K.mp4 golden sample need a special colorimetry "bt2020"
+    # that needs special pipeline configuration to get it streaming smoothly.
+    if golden_sample_path.endswith("4K.mp4"):
+        decoder = (
+            "capssetter replace=true "
+            'caps="video/x-h264, stream-format=(string)byte-stream, '
+            "alignment=(string)au, level=(string)5.2, profile=(string)main, "
+            "width=(int)3840, height=(int)2160, framerate=(fraction)24/1, "
+            "pixel-aspect-ratio=(fraction)1/1, colorimetry=(string)bt2020, "
+            "chroma-format=(string)4:2:0, bit-depth-luma=(uint)8, "
+            'bit-depth-chroma=(uint)8, parsed=(boolean)true" ! {}'
+        ).format(decoder)
```

Comment:
Golden test file is from the customer, please see the [reason](https://bugs.launchpad.net/baoshan/+bug/2073177/comments/1) here.

By the way, I couldn't find a file that is H.264 and 4K resolution on the internet. All available 4k files are HEVC (H.265).

Reply:
are those parameters only fit for this platform? or it's only for this customer platform or the video provided by this customer?

In my opinion, we should move them into configuration file, as we would like to keep this scripts as generic as possible.

Reply:
Indeed, I refactored it, please see https://github.com/canonical/checkbox/pull/1348/commits/9e9725b279089309113aa47ec61bc2ced2fdb079